src/deluge/gui/ui/keyboard/layout/velocity_drums.cpp

```diff
@@ -24,121 +24,191 @@
 namespace deluge::gui::ui::keyboard::layout {
 
 void KeyboardLayoutVelocityDrums::evaluatePads(PressedPad presses[kMaxNumKeyboardPadPresses]) {
-	uint8_t noteIdx = 0;
-
 	currentNotesState = NotesState{}; // Erase active notes
 
 	static_assert(kMaxNumActiveNotes < 32, "We use a 32-bit integer to represent note active state");
-	uint32_t activeNotes = 0;
-	std::array<int32_t, kMaxNumActiveNotes> noteOnTimes{};
-
-	for (int32_t idxPress = 0; idxPress < kMaxNumKeyboardPadPresses; ++idxPress) {
-		if (presses[idxPress].active && presses[idxPress].x < kDisplayWidth) {
-			uint8_t note = noteFromCoords(presses[idxPress].x, presses[idxPress].y);
-			uint8_t velocity = (intensityFromCoords(presses[idxPress].x, presses[idxPress].y) >> 1);
-			auto noteOnIdx = currentNotesState.enableNote(note, velocity);
-
-			// exceeded maximum number of active notes, ignore this note-on
-			if (noteOnIdx == kMaxNumActiveNotes) {
-				continue;
-			}
-
-			if ((activeNotes & (1 << noteOnIdx)) == 0) {
-				activeNotes |= 1 << noteOnIdx;
-				noteOnTimes[noteOnIdx] = presses[idxPress].timeLastPadPress;
-			}
-			else {
-				// this is a retrigger on the same note, see if we should update the velocity
-				auto lastOnTime = noteOnTimes[noteOnIdx];
-				auto thisOnTime = presses[idxPress].timeLastPadPress;
-				if (util::infinite_a_lt_b(lastOnTime, thisOnTime)) {
-					// this press is more recent, use its velocity.
-					currentNotesState.notes[noteOnIdx].velocity = velocity;
-					noteOnTimes[noteOnIdx] = thisOnTime;
+	uint32_t active_notes = 0;
+	std::array<int32_t, kMaxNumActiveNotes> note_on_times{};
+	uint8_t edge_size_x = (uint32_t)getState().drums.edge_size_x;
+	uint8_t edge_size_y = (uint32_t)getState().drums.edge_size_y;
+	uint8_t offset = getState().drums.scroll_offset;
+	uint8_t pads_per_row = kDisplayWidth / edge_size_x;
+	uint8_t highest_clip_note = getHighestClipNote();
+	bool odd_pad = (edge_size_x % 2 == 1 && edge_size_x > 1);
+	for (int32_t idx_press = 0; idx_press < kMaxNumKeyboardPadPresses; ++idx_press) {
+		if (presses[idx_press].active) {
+			uint8_t x = presses[idx_press].x;
+			if (x < kDisplayWidth) {
```

Comment:
These checks might be easier to read if they're written like below:

```suggestion
		if (!presses[idx_press].active) { continue; }
		uint8_t x = presses[idx_press].x;
		if (x >= kkDisplayWidth) { continue; }
```

That way, anyone reading the loop knows that these checks are more like filters on the presses and not something where interesting things happen off-screen (below the end of the then-block).